VisualRecognitionApi/src/main/java/javax/visrec/spi/ServiceProvider.java

```diff
@@ -0,0 +1,126 @@
+
+    protected ServiceProvider() {
+        // Prevent instantiation, only allowed by subclasses.
+    }
+
+    /**
+     * If multiple implementations of the {@link ServiceProvider} are found on the classpath, then the
+     * {@link ServiceProvider} with the highest value of priority will be used by default.
+     * @return The priority (default = 0(
+     */
+    public int getPriority() {
+        return 0;
+    }
+
+    /**
+     * Get the {@link BuilderService}
+     * @return builder service.
+     */
+    public abstract BuilderService getBuilderService();
+
+    /**
+     * Get the {@link ClassifierService}
+     * @return classifier service
+     */
+    public abstract ClassifierService getClassifierService();
+
+    /**
+     * Get the current {@link ServiceProvider}
+     * @return service provider.
+     * @throws IllegalStateException If there are no service providers found.
+     */
+    public static ServiceProvider current() {
+        final ServiceProvider p = getProviders().get(0);
+        if (Objects.isNull(p)) {
+            throw new IllegalStateException("No ServiceProvider found");
+        }
+        return p;
+    }
+
+    /**
+     * Set the current {@link ServiceProvider}
+     * @param provider The {@link ServiceProvider} to be set as current.
+     * @throws IllegalStateException If there are no service providers found.
+     * @throws IllegalArgumentException If the {@link ServiceProvider} given by the parameters is not known
+     * in the existing list of providers.
+     */
+    public static void setCurrent(final ServiceProvider provider) {
+        Objects.requireNonNull(provider);
+
+        synchronized (LOCK) {
+            final List<ServiceProvider> foundProviders = getProviders();
+            if (foundProviders.isEmpty()) {
+                throw new IllegalStateException("No providers found.");
+            }
+            if (!foundProviders.contains(provider)) {
+                throw new IllegalArgumentException("ServiceProvider given through the parameters is not known.");
+            }
+
+            // Copying list, removes the provider from the arguments and prepends it upfront
+            // on the copied list.
+            final List<ServiceProvider> copiedProviders = new LinkedList<>();
+            Collections.copy(copiedProviders, foundProviders);
+            copiedProviders.remove(provider);
+            copiedProviders.add(0, provider);
+
+            // Make the list unmodifiable to prevent illegal modification
+            providers = Collections.unmodifiableList(copiedProviders);
+        }
+    }
+
+    /**
+     * Gets a list of all available {@link ServiceProvider}s.
+     * @return service providers.
+     */
+    public static List<ServiceProvider> available() {
+        return getProviders();
+    }
+
+    private static List<ServiceProvider> getProviders() {
+        if (Objects.isNull(providers)) {
+            synchronized(LOCK) {
```

Comment:
If there is no other alternative, that is fine. But I do want to point out that synchronized blocks become bottlenecks in multithreaded systems.

Reply:
The synchronization only happens once for every ClassLoader. Is that acceptable? Or perhaps I could move the initialization of the ServiceProvider into an inner enum which is thread-safe.